sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/MultimapUserState.java

```diff
@@ -235,6 +254,122 @@ public K next() {
+      return PrefetchableIterables.concat(
+          Iterables.concat(
+              Iterables.transform(
+                  pendingAddsNow.entrySet(),
+                  entry ->
+                      Iterables.transform(
+                          entry.getValue().getValue(),
+                          value -> Maps.immutableEntry(entry.getValue().getKey(), value)))));
+    }
+
+    Set<Object> pendingRemovesNow = new HashSet<>(pendingRemoves.keySet());
+    // Make a deep copy of pendingAdds so this iterator represents a snapshot of state at the time
+    // it was created.
+    Map<K, List<V>> pendingAddsNow = new HashMap<>();
+    for (Map.Entry<Object, KV<K, List<V>>> entry : pendingAdds.entrySet()) {
+      pendingAddsNow.put(
+          entry.getValue().getKey(), new ArrayList<>()); // entry.getValue().getValue());
+      for (V value : entry.getValue().getValue()) {
+        List<V> values = pendingAddsNow.get(entry.getValue().getKey());
+        values.add(value);
+      }
+    }
+    return new PrefetchableIterables.Default<Map.Entry<K, V>>() {
+      @Override
+      public PrefetchableIterator<Map.Entry<K, V>> createIterator() {
+        return new PrefetchableIterator<Map.Entry<K, V>>() {
+          PrefetchableIterator<Map.Entry<K, V>> persistedEntriesIterator =
+              PrefetchableIterables.concat(
+                      Iterables.concat(
+                          Iterables.transform(
+                              persistedEntries,
+                              entry ->
+                                  Iterables.transform(
+                                      entry.getValue(),
+                                      value -> Maps.immutableEntry(entry.getKey(), value)))))
+                  .iterator();
+          Iterator<Map.Entry<K, V>> pendingAddsNowIterator;
+          boolean hasNext;
+          Map.Entry<K, V> nextEntry;
+
+          @Override
+          public boolean isReady() {
+            return persistedEntriesIterator.isReady();
+          }
+
+          @Override
+          public void prefetch() {
+            if (!isReady()) {
+              persistedEntriesIterator.prefetch();
+            }
+          }
+
+          @Override
+          public boolean hasNext() {
+            if (hasNext) {
+              return true;
+            }
+
+            while (persistedEntriesIterator.hasNext()) {
+              Map.Entry<K, V> nextPersistedEntry = persistedEntriesIterator.next();
+              Object nextKeyStructuralValue =
+                  mapKeyCoder.structuralValue(nextPersistedEntry.getKey());
+              if (!pendingRemovesNow.contains(nextKeyStructuralValue)) {
+                nextEntry =
+                    Maps.immutableEntry(nextPersistedEntry.getKey(), nextPersistedEntry.getValue());
+                hasNext = true;
+                return true;
+              }
+            }
+
+            if (pendingAddsNowIterator == null) {
```

Comment:
do we need to merge the persisted and local adds? Does the iterator have any guarantees on ordering of elements for a key?